<a href="https://colab.research.google.com/github/wojr21/courses/blob/main/DS_bootcamp_exercises/141_150_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.18.5'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [2]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))

df = pd.DataFrame({'col1': s1, 'col2': s2})
df

col1      col2
0   0.374540 -1.012831
1   0.950714  0.314247
2   0.731994 -0.908024
3   0.598658 -1.412304
4   0.156019  1.465649
5   0.155995 -0.225776
6   0.058084  0.067528
7   0.866176 -1.424748
8   0.601115 -0.544383
9   0.708073  0.110923
10  0.020584 -1.150994
11  0.969910  0.375698
12  0.832443 -0.600639
13  0.212339 -0.291694
14  0.181825 -0.601707
15  0.183405  1.852278
16  0.304242 -0.013497
17  0.524756 -1.057711
18  0.431945  0.822545
19  0.291229 -1.220844

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [3]:
df[(df['col2'] >= 0) & (df['col2'] <= 1.0)]

col1      col2
1   0.950714  0.314247
6   0.058084  0.067528
9   0.708073  0.110923
11  0.969910  0.375698
18  0.431945  0.822545

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [6]:
df['col3'] = df['col2'].map(lambda x: 1 if x >= 0 else -1)
df

col1      col2  col3
0   0.374540 -1.012831    -1
1   0.950714  0.314247     1
2   0.731994 -0.908024    -1
3   0.598658 -1.412304    -1
4   0.156019  1.465649     1
5   0.155995 -0.225776    -1
6   0.058084  0.067528     1
7   0.866176 -1.424748    -1
8   0.601115 -0.544383    -1
9   0.708073  0.110923     1
10  0.020584 -1.150994    -1
11  0.969910  0.375698     1
12  0.832443 -0.600639    -1
13  0.212339 -0.291694    -1
14  0.181825 -0.601707    -1
15  0.183405  1.852278     1
16  0.304242 -0.013497    -1
17  0.524756 -1.057711    -1
18  0.431945  0.822545     1
19  0.291229 -1.220844    -1

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$. 
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [7]:
df['col4'] = df['col2'].clip(-1.0, 1.0)
df

col1      col2  col3      col4
0   0.374540 -1.012831    -1 -1.000000
1   0.950714  0.314247     1  0.314247
2   0.731994 -0.908024    -1 -0.908024
3   0.598658 -1.412304    -1 -1.000000
4   0.156019  1.465649     1  1.000000
5   0.155995 -0.225776    -1 -0.225776
6   0.058084  0.067528     1  0.067528
7   0.866176 -1.424748    -1 -1.000000
8   0.601115 -0.544383    -1 -0.544383
9   0.708073  0.110923     1  0.110923
10  0.020584 -1.150994    -1 -1.000000
11  0.969910  0.375698     1  0.375698
12  0.832443 -0.600639    -1 -0.600639
13  0.212339 -0.291694    -1 -0.291694
14  0.181825 -0.601707    -1 -0.601707
15  0.183405  1.852278     1  1.000000
16  0.304242 -0.013497    -1 -0.013497
17  0.524756 -1.057711    -1 -1.000000
18  0.431945  0.822545     1  0.822545
19  0.291229 -1.220844    -1 -1.000000

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [10]:
df['col2'].nlargest(5)

15    1.852278
4     1.465649
18    0.822545
11    0.375698
1     0.314247
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [11]:
df['col2'].nsmallest(5)

7    -1.424748
3    -1.412304
19   -1.220844
10   -1.150994
17   -1.057711
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [12]:
df.cumsum()

col1      col2  col3      col4
0   0.374540 -1.012831    -1 -1.000000
1   1.325254 -0.698584     0 -0.685753
2   2.057248 -1.606608    -1 -1.593777
3   2.655907 -3.018912    -2 -2.593777
4   2.811925 -1.553263    -1 -1.593777
5   2.967920 -1.779039    -2 -1.819553
6   3.026004 -1.711511    -1 -1.752025
7   3.892180 -3.136259    -2 -2.752025
8   4.493295 -3.680642    -3 -3.296408
9   5.201367 -3.569719    -2 -3.185485
10  5.221952 -4.720713    -3 -4.185485
11  6.191862 -4.345015    -2 -3.809787
12  7.024304 -4.945653    -3 -4.410426
13  7.236643 -5.237347    -4 -4.702119
14  7.418468 -5.839054    -5 -5.303826
15  7.601873 -3.986776    -4 -4.303826
16  7.906115 -4.000273    -5 -4.317323
17  8.430872 -5.057984    -6 -5.317323
18  8.862817 -4.235439    -5 -4.494778
19  9.154046 -5.456283    -6 -5.494778

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [13]:
df['col2'].median()

-0.4180382371592297

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [14]:
df[df['col2'] > 0.0]

col1      col2  col3      col4
1   0.950714  0.314247     1  0.314247
4   0.156019  1.465649     1  1.000000
6   0.058084  0.067528     1  0.067528
9   0.708073  0.110923     1  0.110923
11  0.969910  0.375698     1  0.375698
15  0.183405  1.852278     1  1.000000
18  0.431945  0.822545     1  0.822545

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [15]:
df[:5].to_dict()

{'col1': {0: 0.3745401188473625,
  1: 0.9507143064099162,
  2: 0.7319939418114051,
  3: 0.5986584841970366,
  4: 0.15601864044243652},
 'col2': {0: -1.0128311203344238,
  1: 0.3142473325952739,
  2: -0.9080240755212109,
  3: -1.4123037013352915,
  4: 1.465648768921554},
 'col3': {0: -1, 1: 1, 2: -1, 3: -1, 4: 1},
 'col4': {0: -1.0,
  1: 0.3142473325952739,
  2: -0.9080240755212109,
  3: -1.0,
  4: 1.0}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [21]:
df_markdown = df.head().to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [23]:
print(df_markdown)

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.37454  | -1.01283  |     -1 | -1        |
|  1 | 0.950714 |  0.314247 |      1 |  0.314247 |
|  2 | 0.731994 | -0.908024 |     -1 | -0.908024 |
|  3 | 0.598658 | -1.4123   |     -1 | -1        |
|  4 | 0.156019 |  1.46565  |      1 |  1        |


Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.37454  | -1.01283  |     -1 | -1        |
|  1 | 0.950714 |  0.314247 |      1 |  0.314247 |
|  2 | 0.731994 | -0.908024 |     -1 | -0.908024 |
|  3 | 0.598658 | -1.4123   |     -1 | -1        |
|  4 | 0.156019 |  1.46565  |      1 |  1        |